### ISSUE TITLE
Support zLinux(s390x) platform

### ISSUE THREAD

**Issue by dpkulkarni** [ManageIQ/manageiq-rpm_build#83]
Hello 

We are trying to port `ManageIQ` on zLinux(s390x) architecture. 

I have started to look into the repos to build images to support s390x platform. There are some rpms/packages not available for s390x. 

**Motivation**:
Red Hat Enterprise Linux 8 is already supporting s390x (IBM Z ). We want to port ManageIQ on zLinux architecture

**Issue**: 
Errors during downloading metadata for repository 'manageiq-11-kasparov':

**Logs:**
```
# docker build -t new_rpm:latest -f Dockerfile .
Emulate Docker CLI using podman. Create /etc/containers/nodocker to quiet msg.
STEP 1: FROM registry.access.redhat.com/ubi8/ubi:8.2
STEP 2: ARG ARCH=s390x
--> Using cache 6b3f69e3e53c526132177da92c581b8740b219b4216b53f491a93bf247f6c751
STEP 3: ENV TERM=xterm     APPLIANCE=true     RAILS_USE_MEMORY_STORE=true
--> Using cache 7dbb5f4bf716965b11f2e65e0b7eab275f9ac0be317d58085a03493d73c29949
STEP 4: RUN dnf -y update
--> Using cache c76d579a23c608abff4c5598f005926a1a164a140202d715c056dd8fae35bfb6
STEP 5: RUN curl -L https://releases.ansible.com/ansible-runner/ansible-runner.el8.repo > /etc/yum.repos.d/ansible-runner.repo
--> Using cache 3e18158aac568cba6fe3173533b17b40ed2bb92d143d1b2e378498a5a2f910f7
STEP 6: RUN dnf install -y https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm
--> Using cache 21afc3244f3a1b7d40dbafad0f526d5133cc8906bd3e5ec3aab10755ec369059
STEP 7: RUN dnf install -y  https://rpm.manageiq.org/release/11-kasparov/el8/noarch/manageiq-release-11.0-1.el8.noarch.rpm &&     dnf -y --disableplugin=subscription-manager module enable ruby:2.6 &&     dnf -y --disableplugin=subscription-manager module enable nodejs:12 &&     dnf -y --disableplugin=subscription-manager groupinstall "development tools" &&     dnf config-manager --setopt=epel.exclude=*qpid-proton* --save &&     dnf -y --disableplugin=subscription-manager --setopt=tsflags=nodocs install       ansible       cmake       copr-cli       createrepo       glibc-langpack-en       libcurl-devel       libpq-devel       libxml2-devel       libxslt-devel       nodejs       openssl-devel       platform-python-devel       postgresql-server       postgresql-server-devel       ruby-devel       rubygem-bundler       sqlite-devel       wget
Updating Subscription Management repositories.
Unable to read consumer identity
Subscription Manager is operating in container mode.
Extra Packages for Enterprise Linux Modular 8 -  49 kB/s |  82 kB     00:01
Extra Packages for Enterprise Linux 8 - s390x   419 kB/s | 6.7 MB     00:16
Red Hat Enterprise Linux 8 for IBM z Systems -  4.3 kB/s | 2.8 kB     00:00
Red Hat Enterprise Linux 8 for IBM z Systems -  3.5 kB/s | 2.4 kB     00:00
manageiq-release-11.0-1.el8.noarch.rpm           33 kB/s |  12 kB     00:00
Dependencies resolved.
================================================================================
 Package                Architecture Version           Repository          Size
================================================================================
Installing:
 manageiq-release       noarch       11.0-1.el8        @commandline        12 k

Transaction Summary
================================================================================
Install  1 Package

Total size: 12 k
Installed size: 7.5 k
Downloading Packages:
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                        1/1
  Installing       : manageiq-release-11.0-1.el8.noarch                     1/1
  Verifying        : manageiq-release-11.0-1.el8.noarch                     1/1
Installed products updated.

Installed:
  manageiq-release-11.0-1.el8.noarch

Complete!
ManageIQ 11 (Kasparov) Release- s390x           339  B/s | 224  B     00:00
Errors during downloading metadata for repository 'manageiq-11-kasparov':
  - Status code: 404 for https://rpm.manageiq.org/release/11-kasparov/el8/s390x/repodata/repomd.xml (IP: 205.185.216.10)
Error: Failed to download metadata for repo 'manageiq-11-kasparov': Cannot download repomd.xml: Cannot download repodata/repomd.xml: All mirrors were tried
Error: error building at STEP "RUN dnf install -y  https://rpm.manageiq.org/release/11-kasparov/el8/noarch/manageiq-release-11.0-1.el8.noarch.rpm &&     dnf -y --disableplugin=subscription-manager module enable ruby:2.6 &&     dnf -y --disableplugin=subscription-manager module enable nodejs:12 &&     dnf -y --disableplugin=subscription-manager groupinstall "development tools" &&     dnf config-manager --setopt=epel.exclude=*qpid-proton* --save &&     dnf -y --disableplugin=subscription-manager --setopt=tsflags=nodocs install       ansible       cmake       copr-cli       createrepo       glibc-langpack-en       libcurl-devel       libpq-devel       libxml2-devel       libxslt-devel       nodejs       openssl-devel       platform-python-devel       postgresql-server       postgresql-server-devel       ruby-devel       rubygem-bundler       sqlite-devel       wget": error while running runtime: exit status 1
#
```

**Environment:**
```
# uname -a
Linux adoring1.fyre.ibm.com 4.18.0-193.el8.s390x #1 SMP Fri Mar 27 14:43:09 UTC 2020 s390x s390x s390x GNU/Linux
# docker --version
podman version 1.4.2-stable3
# cat /etc/os-release
NAME="Red Hat Enterprise Linux"
VERSION="8.2 (Ootpa)"
ID="rhel"
ID_LIKE="fedora"
VERSION_ID="8.2"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Red Hat Enterprise Linux 8.2 (Ootpa)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:redhat:enterprise_linux:8.2:GA"
HOME_URL="https://www.redhat.com/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"

REDHAT_BUGZILLA_PRODUCT="Red Hat Enterprise Linux 8"
REDHAT_BUGZILLA_PRODUCT_VERSION=8.2
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux"
REDHAT_SUPPORT_PRODUCT_VERSION="8.2"
```

JFYI - Travis CI is also supporting s390x environment - which makes development on s390x easier.
https://blog.travis-ci.com/2019-11-12-multi-cpu-architecture-ibm-power-ibm-z

Please let us know how do we collaborate and contribute back to ManageIQ

Thanks

**Comment by dpkulkarni** [comment]
@chessbyte 

I am further able to proceed ahead building the `manageiq-rpm_build` rpms for s390x.
```
# find . -name *.rpm
./BUILD/rpms/s390x/manageiq-appliance-11.0.0-0.1.20200826101332.el8.s390x.rpm
./BUILD/rpms/s390x/manageiq-appliance-tools-11.0.0-0.1.20200826101332.el8.s390x.rpm
./BUILD/rpms/s390x/manageiq-core-11.0.0-0.1.20200826101332.el8.s390x.rpm
./BUILD/rpms/s390x/manageiq-gemset-11.0.0-0.1.20200826101332.el8.s390x.rpm
./BUILD/rpms/s390x/manageiq-pods-11.0.0-0.1.20200826101332.el8.s390x.rpm
./BUILD/rpms/s390x/manageiq-system-11.0.0-0.1.20200826101332.el8.s390x.rpm
./BUILD/rpms/s390x/manageiq-ui-11.0.0-0.1.20200826101332.el8.s390x.rpm
```
There were interim failures for which I had to add many dependencies for s390x which I am listing below: 
```
yum -y install ruby-devel gcc make 
gem install bundler
dnf install ruby-devel cmake mariadb-devel rubygems-devel
bundler update
bundler install
yum install postgresql-devel 
yum install gcc-c++
gem install sassc -- --disable-march-tune-native
gem install unf_ext -v '0.0.7.2' -- --with-cxxflags=-fsigned-char
yum install libxml2-devel
yum install libcurl-devel
 dnf -y install https://kojipkgs.fedoraproject.org/packages/qpid-proton/0.30.0/2.el8/s390x/qpid-proton-c-0.30.0-2.el8.s390x.rpm
 dnf -y install https://kojipkgs.fedoraproject.org/packages/qpid-proton/0.30.0/2.el8/s390x/qpid-proton-c-devel-0.30.0-2.el8.s390x.rpm
  yum install nodejs
curl --silent --location https://dl.yarnpkg.com/rpm/yarn.repo | sudo tee /etc/yum.repos.d/yarn.repo
yum install -y yarn
install node > 1.12 // wget https://nodejs.org/dist/v12.18.3/node-v12.18.3-linux-s390x.tar.xz untar 
yum install python36 // may not be required
yum install platform-python-devel  //pathfix.py
```
The above list is not properly arranged which I will do while raising PR. 

I was facing one issue in `/root/manageiq-rpm_build/lib/manageiq/rpm_build/generate_core.rb` file with the below error:
![image](https://user-images.githubusercontent.com/64318450/91381971-461ecd80-e846-11ea-991f-95307b093cf7.png)

For now I commented out the `rake path` check and it proceeded. Could you please guide me on this issue to resolve.

/cc @simaishi @Fryguy

**Comment by simaishi** [comment]
I will add s390x qpid-proton to our yum repo, which will resolve your original issue of not finding manageiq-11-kasparov for s390x. And once that's done, you can get rid of installing qpid-proton from koji.

I don't see CentOS 8 supporting s390x. But since you were able to build manageiq RPMs, I assume you were able to get all needed packages from ubi + epel. Not sure what's listed above is all packages installed or not... For example, you need libssh2-devel on the build machine. The build won't fail even without it, but a feature will be missing if built without it. We can review the list once cleaned up.

As for the rake path, it's hard to say without looking at all the changes you've made to get to that point. But I think I've seen that error when some of the gem environment variable (like GEM_PATH or GEM_HOME) wasn't set up the way it was supposed to be.

**Comment by simaishi** [comment]
> I will add s390x qpid-proton to our yum repo, which will resolve your original issue of not finding manageiq-11-kasparov for s390x. 

Done.

**Comment by dpkulkarni** [comment]
@simaishi I am able to build `manageiq-rpm_base` image for zLinux. 
```
# docker images | grep rpm
localhost/manageiq-rpm_build          latest   388021a1e8fb   3 hours ago      1.38 GB
```
Now I am trying to build image for manageiq-base as this is required base image for the next consecutive docker images.

I am trying to get docker image with this command:
`./bin/build -d images/ -r manageiq-base`

I am getting the below error- 

![image](https://user-images.githubusercontent.com/64318450/91982533-04ab8800-ed48-11ea-858e-2b58566afa71.png)

How can I get the `manageiq-pods` unless I build the manageiq-base image. could you please help me to understand.

**Comment by simaishi** [comment]
@dpkulkarni Since we don't build s390x manageiq-* rpms, you have to build rpms using the locally built manageiq-rpm_build image. You mentioned you were able to build rpms at one point, so I assume that's still the case.

manageiq-pods bin/build script has an option to build rpms locally before building container images:

`RPM_BUILD_IMAGE=localhost/manageiq-rpm_build RPM_BUILD_OPTIONS=<path to override options.yml> bin/build -d images -r manageiq -b`

If you don't need to override anything in options.yml (either because you're building with default values or your local rpm_build image already has all options specified), you don't need to use RPM_BUILD_OPTIONS.

If you already built rpms and have them available, you can copy s390x directory to images/manageiq-base and run:

`bin/build -d images -r manageiq -l`

**Comment by dpkulkarni** [comment]
@simaishi Thanks for your pointers I am able to proceed ahead. 

I am facing issue in building the rpms using this command:
`RPM_BUILD_IMAGE=localhost/manageiq/rpm_build:latest bin/build -d images -r manageiq -b`
I am getting **Permission denied** error for s390x. 

```
    Duplicate build-ids /root/rpmbuild/BUILDROOT/manageiq-11.0.0-0.1.20200904204337.el8.s390x/opt/manageiq/manageiq-gemset/extensions/s390x-linux/2.6.0/qpid_proton-0.30.0/cproton.so and /root/rpmbuild/BUILDROOT/manageiq-11.0.0-0.1.20200904204337.el8.s390x/opt/manageiq/manageiq-gemset/gems/qpid_proton-0.30.0/ext/cproton/cproton.so
    Duplicate build-ids /root/rpmbuild/BUILDROOT/manageiq-11.0.0-0.1.20200904204337.el8.s390x/opt/manageiq/manageiq-gemset/extensions/s390x-linux/2.6.0/sqlite3-1.3.13/sqlite3/sqlite3_native.so and /root/rpmbuild/BUILDROOT/manageiq-11.0.0-0.1.20200904204337.el8.s390x/opt/manageiq/manageiq-gemset/gems/sqlite3-1.3.13/ext/sqlite3/sqlite3_native.so
    cannot create /root/BUILD/rpms/s390x: Permission denied
    Could not open /root/BUILD/rpms/s390x/manageiq-appliance-11.0.0-0.1.20200904204337.el8.s390x.rpm: No such file or directory
#
```
I am working as root so why I am getting this error.

**Comment by dpkulkarni** [comment]
@simaishi need your help to identify the `Permission denied` issue.  Is that I am missing out something in configuration.?

**Comment by dpkulkarni** [comment]
@simaishi 

As you suggested - 
I tried another approach of copying already built `rpms` to the  `images/manageiq-base/rpms` directory as below:
```
# ls images/manageiq-base/rpms/s390x/
manageiq-appliance-11.0.0-0.1.20200907030356.el8.s390x.rpm        manageiq-pods-11.0.0-0.1.20200907030356.el8.s390x.rpm
manageiq-appliance-tools-11.0.0-0.1.20200907030356.el8.s390x.rpm  manageiq-system-11.0.0-0.1.20200907030356.el8.s390x.rpm
manageiq-core-11.0.0-0.1.20200907030356.el8.s390x.rpm             manageiq-ui-11.0.0-0.1.20200907030356.el8.s390x.rpm
manageiq-gemset-11.0.0-0.1.20200907030356.el8.s390x.rpm
```
then I tried building images using the local rpms, using the below command:
`RPM_BUILD_IMAGE=localhost/manageiq/rpm_build:latest bin/build -d images -r manageiq -l true`
but it failed with the below error: 
```

Complete!
Updating Subscription Management repositories.
Unable to read consumer identity
Subscription Manager is operating in container mode.
Local yum repo                                                                                                257 MB/s | 324 kB     00:00
ManageIQ 11 (Kasparov) Release- s390x                                                                         2.4 kB/s | 3.0 kB     00:01
ManageIQ 11 (Kasparov) Release- noarch                                                                         15 kB/s | 3.0 kB     00:00
ManageIQ 11 (Kasparov) Nightly- s390x                                                                          16 kB/s | 3.0 kB     00:00
Red Hat Enterprise Linux 8 for IBM z Systems - AppStream (RPMs)                                               4.2 kB/s | 2.8 kB     00:00
Red Hat Enterprise Linux 8 for IBM z Systems - BaseOS (RPMs)                                                  3.6 kB/s | 2.4 kB     00:00
Error:
 Problem: package manageiq-core-11.0.0-0.1.20200907030356.el8.s390x requires manageiq-gemset = 11.0.0-0.1.20200907030356.el8, but none of the providers can be installed
  - package manageiq-pods-11.0.0-0.1.20200907030356.el8.s390x requires manageiq-core = 11.0.0-0.1.20200907030356.el8, but none of the providers can be installed
  - package manageiq-gemset-11.0.0-0.1.20200907030356.el8.s390x requires libruby.so.2.5()(64bit), but none of the providers can be installed
  - conflicting requests
  - package ruby-libs-2.5.3-104.module+el8.0.0+3250+4b7d6d43.s390x is filtered out by modular filtering
  - package ruby-libs-2.5.5-105.module+el8.1.0+3656+f80bfa1d.s390x is filtered out by modular filtering
  - package ruby-libs-2.5.3-103.module+el8+2671+ebcc7ee0.s390x is filtered out by modular filtering
(try to add '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)
Error: error building at STEP "RUN dnf -y --disableplugin=subscription-manager install       https://dl.fedoraproject.org/pub/epel/epel-release-latest-8.noarch.rpm       https://rpm.manageiq.org/release/11-kasparov/el8/noarch/manageiq-release-11.0-1.el8.noarch.rpm       https://resources.ovirt.org/pub/yum-repo/ovirt-release44.rpm &&     dnf -y --disableplugin=subscription-manager module enable nodejs:12 &&     dnf -y --disableplugin=subscription-manager module enable ruby:2.6 &&     if [[ "$RELEASE_BUILD" != "true" ]]; then dnf config-manager --enable manageiq-11-kasparov-nightly; fi &&     dnf config-manager --setopt=ubi-8-appstream.exclude=*net-snmp* --setopt=ubi-8-baseos.exclude=*net-snmp* --setopt=ubi-8-codeready-builder.exclude=*net-snmp* --save &&     if [[ "$LOCAL_RPM" = "true" ]]; then /create_local_yum_repo.sh; fi &&     dnf -y --disableplugin=subscription-manager --setopt=tsflags=nodocs install       ${RPM_PREFIX}-pods                python3-devel                     &&                              dnf -y --disableplugin=subscription-manager upgrade &&     dnf clean all": error while running runtime: exit status 1
[root@scare1 manageiq-pods]#
```

Here is the listing of rpm_sepc dir:
```
# ls -ltr /root/BUILD/rpm_spec
total 101476
drwxr-xr-x. 2 root root      171 Sep  7 03:03 subpackages
-rw-r--r--. 1 root root     4144 Sep  7 03:03 manageiq.spec.in
-rw-r--r--. 1 root root       93 Sep  7 03:03 changelog
-rw-r--r--. 1 root root 45959826 Sep  7 03:24 manageiq-core-11.0.0.tar.gz
-rw-r--r--. 1 root root 57863243 Sep  7 03:25 manageiq-gemset-11.0.0.tar.gz
-rw-r--r--. 1 root root    32519 Sep  7 03:25 manageiq-appliance-11.0.0.tar.gz
-rw-r--r--. 1 root root    24806 Sep  7 03:25 manageiq-manifest-11.0.0.tar.gz
-rw-r--r--. 1 root root    10565 Sep  7 03:25 manageiq.spec
```

Need your guidance to build the images.

**Comment by simaishi** [comment]
@dpkulkarni Since there is another issue opened for building pods with local rpms (https://github.com/ManageIQ/manageiq-pods/issues/622), let's focus on problems building rpms in this particular issue. Is there any outstanding issue building rpms?

**Comment by dpkulkarni** [comment]
Yes @simaishi I still have this issue while building the rpms as mentioned above (https://github.com/ManageIQ/manageiq-rpm_build/issues/83#issuecomment-687388664). Do you have any idea on the `Permission denied` ?

**Comment by simaishi** [comment]
Can you verify mounting volume works in your environment? There have been a few issues due to differences in environment, so it's a good idea to make sure you can mount volumes and write to it.

**Comment by dpkulkarni** [comment]
Able to build the rpms using the "-b" option. 
```
# ls BUILD/rpms/s390x/*
BUILD/rpms/s390x/manageiq-appliance-11.0.0-0.1.20200907030356.el8.s390x.rpm
BUILD/rpms/s390x/manageiq-appliance-tools-11.0.0-0.1.20200907030356.el8.s390x.rpm
BUILD/rpms/s390x/manageiq-core-11.0.0-0.1.20200907030356.el8.s390x.rpm
BUILD/rpms/s390x/manageiq-gemset-11.0.0-0.1.20200907030356.el8.s390x.rpm
BUILD/rpms/s390x/manageiq-pods-11.0.0-0.1.20200907030356.el8.s390x.rpm
BUILD/rpms/s390x/manageiq-system-11.0.0-0.1.20200907030356.el8.s390x.rpm
BUILD/rpms/s390x/manageiq-ui-11.0.0-0.1.20200907030356.el8.s390x.rpm
```

**Comment by dpkulkarni** [comment]
@simaishi 
Raised PR https://github.com/ManageIQ/manageiq-rpm_build/pull/94

**Comment by dpkulkarni** [comment]
@simaishi attaching the build log for manageiq-rpm_build image
[build_log_manageiq-rpm_build.txt](https://github.com/ManageIQ/manageiq-rpm_build/files/5306025/build_log_manageiq-rpm_build.txt)

**Comment by dpkulkarni** [comment]
@simaishi Please find the build log 
[build_log_manageiq-rpm_build.txt](https://github.com/ManageIQ/manageiq-rpm_build/files/5306278/build_log_manageiq-rpm_build.txt)

**Comment by simaishi** [comment]
@dpkulkarni please add another command to Dockerfile and get me the output:

`dnf module provides libssh2`

You can replace `dnf module list --enabled` with this.

**Comment by dpkulkarni** [comment]
@simaishi please find the build log with `dnf module provides libssh2`
[build_log_manageiq-rpm_build.txt](https://github.com/ManageIQ/manageiq-rpm_build/files/5318953/build_log_manageiq-rpm_build.txt)

**Comment by simaishi** [comment]
Thank you, now I see the difference. In x86_64 build, `module provides libssh2` doesn't return anything. Can you try adding `dnf module disable virt:rhel` and see if makes a difference?

**Comment by dpkulkarni** [comment]
@simaishi 
Built the docker image with `dnf module disable virt:rhel`

**Comment by chessbyte** [comment]
@dpkulkarni is this issue addressed by #94 or are there other PRs needed to address this?

**Comment by dpkulkarni** [comment]
@simaishi thanks for all the support to add s390x arch for ManageIQ.

Thanks @chessbyte we can close this issue
